ddl/ddl_api.go

```diff
@@ -3931,3 +3932,77 @@ func extractCollateFromOption(def *ast.ColumnDef) []string {
 	}
 	return specifiedCollates
 }
+
+func (d *ddl) RepairTable(ctx sessionctx.Context, table *ast.TableName, createStmt *ast.CreateTableStmt) error {
+	// Existence of DB and table has been checked in the preprocessor.
+	oldTableInfo, ok := (ctx.Value(domainutil.RepairedTable)).(*model.TableInfo)
+	if !ok || oldTableInfo == nil {
+		return ErrRepairTableFail.GenWithStack("Failed to get the repaired table")
+	}
+	oldDBInfo, ok := (ctx.Value(domainutil.RepairedDatabase)).(*model.DBInfo)
+	if !ok || oldDBInfo == nil {
+		return ErrRepairTableFail.GenWithStack("Failed to get the repaired database")
+	}
+	// By now only support same DB repair.
+	if createStmt.Table.Schema.L != oldDBInfo.Name.L {
+		return ErrRepairTableFail.GenWithStack("Repaired table should in same database with the old one")
+	}
+	// It is necessary to specify the table.ID and partition.ID manually.
+	newTableInfo, err := buildTableInfoWithCheck(ctx, d, createStmt, oldTableInfo.Charset, oldTableInfo.Collate)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// Override newTableInfo with oldTableInfo's element necessary.
+	// TODO: There may be more element assignments here, and the new TableInfo should be verified with the actual data.
+	newTableInfo.ID = oldTableInfo.ID
+	if err = checkAndOverridePartitionID(newTableInfo, oldTableInfo); err != nil {
+		return err
+	}
+	newTableInfo.AutoIncID = oldTableInfo.AutoIncID
+	// If any old columnInfo has lost, that means the old column ID lost too, repair failed.
+	for i, newOne := range newTableInfo.Columns {
+		old := getColumnInfoByName(oldTableInfo, newOne.Name.L)
+		if old == nil {
+			return ErrRepairTableFail.GenWithStackByArgs("Column " + newOne.Name.L + " has lost")
+		}
+		if newOne.Tp != old.Tp {
+			return ErrRepairTableFail.GenWithStackByArgs("Column " + newOne.Name.L + " type should be the same")
```

Comment:
Could we add the type of column to the error message?
Do we need to check the `Flen` ?

Reply:
We don't assume the too much element in the old version is reasonable，providing a certain degree of freedom for more robust repairs.

Reply:
Addressed.